streams/jvm/src/main/scala/zio/stream/compression/Gunzipper.scala

```
@@ -0,0 +1,181 @@
+package zio.stream.compression
+
+import java.util.zip.{ CRC32, Inflater }
+import java.{ util => ju }
+
+import scala.annotation.tailrec
+
+import zio._
+
+/** Performs few steps of parsing header, then decompresses body and checks trailer.
+ * With reasonably chosen bufferSize there shouldn't occur many concatenation of arrays. */
+private[compression] class Gunzipper private (bufferSize: Int) {
+
+  import Gunzipper._
+
+  private var state: State = new ParseHeaderStep(Array.emptyByteArray, new CRC32)
+
+  def reset(): Unit = {
+    state.close()
+    state = new ParseHeaderStep(Array.emptyByteArray, new CRC32)
+  }
+
+  def close(): Unit = state.close()
+
+  def onChunk(c: Chunk[Byte]): ZIO[Any, CompressionException, Chunk[Byte]] =
+    ZIO.effect {
+      val (newState, output) = state.feed(c.toArray)
+      state = newState
+      output
+    }.refineOrDie {
+      case e: ju.zip.DataFormatException => CompressionException(e)
+      case e: CompressionException       => e
+    }
+
+  private def nextStep(
+    acc: Array[Byte],
+    checkCrc16: Boolean,
+    crc32: CRC32,
+    parseExtra: Boolean,
+    commentsToSkip: Int
+  ): Gunzipper.State =
+    if (parseExtra) new ParseExtraStep(acc, crc32, checkCrc16, commentsToSkip)
+    else if (commentsToSkip > 0) new SkipCommentsStep(checkCrc16, crc32, commentsToSkip)
+    else if (checkCrc16) new CheckCrc16Step(Array.emptyByteArray, crc32.getValue())
+    else new Decompress()
+
+  private class ParseHeaderStep(acc: Array[Byte], crc32: CRC32) extends State {
+
+    //TODO: If whole input is shorther than fixed header, not output is produced and no error is singaled. Is it ok?
```

Sounds like we should throw an exception if the input does not contain a valid header. Is that possible to do?

---

Should I change:
```scala
case None =>
              ZIO.succeed {
                gunzipper.reset()
                Chunk.empty
              }
```
to
```scala
case None =>
              ZIO.succeed {
                if(gunzipper.isFinished()){
                  gunzipper.reset()
                  Chunk.empty
                } else {
                  <fail because stream wasn't complete gzipped>
                }
              }
``` ?

---

I would prefer the error and not the empty chunk :)  Is there an "auto decompress" path that decompresses if the header exists, and leaves the bytes intact if not?

---

If we would like to have such path, then invalid header would have to fallback to clear text stream. Because there is no criterion to distinguish between invalid header and no header.

---

@LGLO To your question: Yes, you should add that `else` branch, but you should use `ZIO.effect` instead.

---

Thanks!
I didn't left this work here. Hopefully I could spend some time today in the evening and next week.

---

Thank you @LGLO!